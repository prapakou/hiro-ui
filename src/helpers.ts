import { useEffect, useState } from "react";
import { Observable } from "rxjs";

export const createObserver = <T>(target: Observable<T>, initial: T) => () => {
  const [value, setValue] = useState(initial);

  useEffect(() => {
    const sub = target.subscribe({
      next: v => setValue(v)
    });

    return () => sub.unsubscribe();
  }, []);

  return value;
};
