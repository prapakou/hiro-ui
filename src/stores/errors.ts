import { BehaviorSubject } from "rxjs";

import { createStateGetter } from "../helpers";

type ErrorState = Error | undefined;

const error$ = new BehaviorSubject<ErrorState>(undefined);

export const errorStore = {
  actions: {
    clearError: () => error$.next(undefined),
    setError: (e: Error) => error$.next(e)
  },
  getters: {
    useError: createStateGetter<ErrorState>(error$)
  }
};
