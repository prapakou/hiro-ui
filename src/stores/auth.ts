import { AuthAccountVertex } from "@hiro-graph/orm-mappings";
import { BehaviorSubject, Subject } from "rxjs";
import { catchError, filter, switchMap, throttleTime } from "rxjs/operators";

import { Auth } from "../auth";
import { createObserver } from "../helpers";
import { Orm } from "../typings";

import { errorStore } from "./errors";

// Types
interface IAuthState {
  token?: string;
  orm?: Orm;
  me?: AuthAccountVertex;
  updated?: Date;
}

interface IAuthCheck {
  authConfig?: any;
  config?: any;
  orm?: any;
}

type AuthKeys = "authConfig" | "config" | "orm";

// Helpers
const getOrSet = (key: AuthKeys, value: any) => {
  if (store.has(key) && !value) {
    return store.get(key);
  }

  store.set(key, value);
  return value;
};

// Stores
const store = new Map<AuthKeys, any>();
const authCheck$ = new Subject<IAuthCheck>();
const auth$ = new BehaviorSubject<IAuthState>({});

let auth;

class LoginError extends Error {
  constructor(...params) {
    super(...params);
    this.name = "LoginError";
  }
}

// Subscriptions
authCheck$
  .pipe(
    filter(v => !!v.authConfig || (!!v.config && !!v.orm)),
    throttleTime(2000),
    switchMap(async v => {
      if (v.orm && v.config) {
        const me = await v.orm.person();
        return { me, orm: v.orm, token: v.config.token, updated: new Date() };
      }

      if (!auth) {
        auth = new Auth(v.authConfig);
      }

      const res = await auth.isLoggedIn();

      if (res.ok) {
        return res;
      }

      const res2 = await auth.login();
      if (res2.ok) {
        window.location.reload();
        return res2;
      } else {
        throw new LoginError("Failed to login!");
      }
    }),
    catchError(err => {
      errorStore.actions.setError(err);
      return [{}];
    })
  )
  .subscribe(auth$);

const ensureLogin = (newAuthConfig?, newConfig?, newOrm?) => {
  const orm = getOrSet("orm", newOrm);
  const config = getOrSet("config", newConfig);
  const authConfig = getOrSet("authConfig", newAuthConfig);

  authCheck$.next({ orm, config, authConfig });
};

export const authStore = {
  actions: {
    ensureLogin
  },
  getters: {
    getToken: () => {
      return new Promise<string>((resolve, reject) => {
        ensureLogin();
        auth$.subscribe({
          error: reject,
          next: v => resolve(v.token)
        });
      });
    },
    useAuth: createObserver<IAuthState>(auth$, {})
  }
};
