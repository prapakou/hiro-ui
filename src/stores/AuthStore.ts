import { Container } from "unstated";

import { Auth } from "../auth";

export interface IAuthStore {
  token?: string;
  orm?: any;
  me?: any;
  updated?: Date;
}

export class AuthStore extends Container<IAuthStore> {
  auth?: Auth;
  state: IAuthStore = {};
  authConfig?: any;
  config?: any;
  orm?: any;
  constructor(authConfig, config, orm) {
    super();
    this.authConfig = authConfig;
    this.config = config;
    this.orm = orm;
  }

  getToken() {
    return this.state.token;
  }

  async ensureLogin() {
    if (this.orm && this.config) {
      this.orm.person().then(me =>
        this.setState({
          me,
          orm: this.orm,
          token: this.config.token,
          updated: new Date()
        })
      );

      return;
    }

    if (!this.authConfig) {
      return "Config required to check login";
    }

    if (!this.auth) {
      this.auth = new Auth(this.authConfig);
    }

    return this.auth.isLoggedIn().then(async res => {
      await this.setState(res);

      if (res.ok) {
        return;
      }

      if (!this.auth) {
        this.auth = new Auth(this.authConfig);
      }

      const res2 = await this.auth.login();

      if (res2.ok) {
        window.location.reload();
      } else {
        return "Failed to login!";
      }
    });
  }
}
