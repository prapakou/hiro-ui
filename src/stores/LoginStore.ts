import { Container } from "unstated";

import { Auth } from "../auth";

export interface ILoginStore {
  token?: string;
  orm?: any;
  me?: any;
  updated?: Date;
}

export class LoginStore extends Container<ILoginStore> {
  auth?: Auth;
  state: ILoginStore = {};
  authConfig?: any;
  config?: any;
  orm?: any;
  constructor(authConfig, config, orm) {
    super();
    this.authConfig = authConfig;
    this.config = config;
    this.orm = orm;
  }

  getToken() {
    return this.state.token;
  }

  ensureLogin() {
    if (this.orm && this.config) {
      this.orm.person().then(me =>
        this.setState({
          me,
          orm: this.orm,
          token: this.config.token,
          updated: new Date()
        })
      );

      return;
    }

    if (!this.authConfig) {
      console.warn("Config required to check login");
      return;
    }

    if (!this.auth) {
      this.auth = new Auth(this.authConfig);
    }

    this.auth.isLoggedIn().then(async res => {
      this.setState(res);

      if (res.ok) {
        return;
      }

      if (!this.auth) {
        this.auth = new Auth(this.authConfig);
      }

      const res2 = await this.auth.login();

      if (res2.ok) {
        window.location.reload();
      } else {
        console.error("Failed to login!");
      }
    });
  }
}
