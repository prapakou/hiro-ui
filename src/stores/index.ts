import { applyMiddleware, combineReducers, compose, createStore } from "redux";
import createSagaMiddleware, { Saga } from "redux-saga";
import { Reducer } from "typesafe-actions";

import { ERROR_NAMESPACE, errorReducer } from "./errors";
import { HIRO_NAMESPACE } from "./constants";
import { AUTH_NAMESPACE } from "./auth/constants";
import { authReducer } from "./auth";

export * from "./errors";

// Combine built-in sagas
function* mainSaga() {
  // Add sagas
}

export function init(
  state?: any,
  reducers?: { [index: string]: Reducer<any, any> },
  sagas?: Saga[]
) {
  // Combine built-in reducers with exeternal ones
  const rootReducer = combineReducers({
    [HIRO_NAMESPACE]: combineReducers({
      [AUTH_NAMESPACE]: authReducer,
      [ERROR_NAMESPACE]: errorReducer
    }),
    ...reducers
  });

  const composeEnhancers =
    (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  const sagaMiddleware = createSagaMiddleware();

  // Create store
  const store = createStore(
    rootReducer,
    state,
    /* preloadedState, */
    composeEnhancers(applyMiddleware(sagaMiddleware))
  );

  // Run built-in saga
  sagaMiddleware.run(mainSaga);

  // Run external sagas
  if (sagas) {
    for (const saga of sagas) {
      sagaMiddleware.run(saga);
    }
  }

  return store;
}
