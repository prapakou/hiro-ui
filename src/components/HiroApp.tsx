import React, { useCallback, useState } from "react";
import { Store } from "redux";
import { BrowserRouter } from "react-router-dom";
import { Container, Loader } from "semantic-ui-react";
import { StoreContext } from "redux-react-hook";

import { init } from "../stores";

import { ErrorBar } from "./ErrorBar";
import { HiroStyle, ThemeOptions } from "./HiroStyle";
import { HiroAuth, HiroAuthConfig } from "./HiroAuth";

type HiroAppProps = {
  ready?: () => void;
  children?: any;
  fluid?: boolean;
  store?: Store;
  auth?: HiroAuthConfig;
} & ThemeOptions;

export const HiroApp = ({
  children,
  ready,
  theme,
  themeVersion,
  fluid,
  store,
  auth
}: HiroAppProps) => {
  const [loading, setLoading] = useState(true);

  let rootStore = store;

  if (!rootStore) {
    rootStore = init();
  }

  const setReady = useCallback(() => {
    if (ready) {
      ready();
    }
    setLoading(false);
  }, [ready]);

  const content = loading ? (
    <Loader active size="huge" content="Loading..." />
  ) : (
    children
  );

  return (
    <BrowserRouter>
      <StoreContext.Provider value={rootStore}>
        <HiroStyle
          theme={theme}
          themeVersion={themeVersion}
          onLoad={setReady}
        />
        <Container fluid={fluid}>
          {auth ? <HiroAuth config={auth}>{content}</HiroAuth> : content}
          <ErrorBar />
        </Container>
      </StoreContext.Provider>
    </BrowserRouter>
  );
};
