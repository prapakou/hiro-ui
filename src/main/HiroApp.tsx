import React, { ReactNode } from "react";
import ReactDOM from "react-dom";
import { SdkConfig, initSdk } from "@hiro-ui/sdk";
import { Reducer } from "typesafe-actions";
import { Saga } from "redux-saga";

import { init } from "../stores";
import { HiroAppRoot, HiroGraphConfig } from "../components";

interface HiroAppParams {
  config?: SdkConfig;
  reducers?: { [index: string]: Reducer<any, any> };
  sagas?: Saga[];
}

export class HiroApp {
  config?: SdkConfig;

  reducers?: { [index: string]: Reducer<any, any> };

  sagas?: Saga[];

  constructor({ config, reducers, sagas }: HiroAppParams = {}) {
    this.config = config;
    this.reducers = reducers;
    this.sagas = sagas;
  }

  render = async (children: ReactNode, target: HTMLElement | null) => {
    const { ready, state, token, graphUrl } = await initSdk(this.config);
    const store = init(state, this.reducers, this.sagas);

    let authConfig: HiroGraphConfig | undefined;

    if (token && graphUrl) {
      authConfig = {
        endpoint: graphUrl,
        token
      };
    }

    ReactDOM.render(
      <HiroAppRoot store={store} auth={authConfig}>
        {children}
      </HiroAppRoot>,
      target,
      ready
    );
  };
}
