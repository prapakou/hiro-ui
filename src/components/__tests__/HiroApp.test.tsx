import React from "react";
import renderer from "react-test-renderer";

import { HiroApp } from "../HiroApp";
import { useErrorDispatch } from "../../stores";

import { getTree } from "./helpers";

test("Renders", () => {
  const component = renderer.create(<HiroApp>Test</HiroApp>);
  const tree = getTree(component);

  expect(tree).toMatchSnapshot();
});

const Error = () => {
  const { setError } = useErrorDispatch();
  setError({ message: "Test error" });

  return <p>Oh no!!</p>;
};

test("Renders with Error", () => {
  const component = renderer.create(
    <HiroApp>
      Test <Error />
    </HiroApp>
  );
  const tree = getTree(component);

  expect(tree).toMatchSnapshot();
});
