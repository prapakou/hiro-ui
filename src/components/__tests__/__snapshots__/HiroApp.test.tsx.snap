// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders 1`] = `
Array [
  <link
    href="https://dtlv35ikt30on.cloudfront.net/latest/default/semantic.min.css"
    onLoad={[Function]}
    rel="stylesheet"
  />,
  <div
    className="ui container"
  >
    Test
  </div>,
]
`;

exports[`Renders with Error 1`] = `
Array [
  <link
    href="https://dtlv35ikt30on.cloudfront.net/latest/default/semantic.min.css"
    onLoad={[Function]}
    rel="stylesheet"
  />,
  <div
    className="ui container"
  >
    Test 
    <p>
      Oh no!!
    </p>
    <div
      className="ui error message"
      style={
        Object {
          "backgroundColor": "mistyrose",
          "bottom": 0,
          "left": 0,
          "position": "fixed",
          "right": 0,
        }
      }
    >
      <div
        className="ui right internal close rail"
      >
        <button
          className="ui red compact icon right floated button"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="close icon"
            onClick={[Function]}
          />
        </button>
      </div>
      <b>
        Error
      </b>
      <p>
        Test error
      </p>
    </div>
  </div>,
]
`;
